homeassistant/components/modern_forms/fan.py

```diff
@@ -0,0 +1,181 @@
+"""Support for Modern Forms Fan Fans."""
+from __future__ import annotations
+
+from functools import partial
+from typing import Any, Callable
+
+from aiomodernforms.const import FAN_POWER_OFF, FAN_POWER_ON
+import voluptuous as vol
+
+from homeassistant.components.fan import SUPPORT_DIRECTION, SUPPORT_SET_SPEED, FanEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import callback
+import homeassistant.helpers.entity_platform as entity_platform
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.util.percentage import (
+    int_states_in_range,
+    percentage_to_ranged_value,
+    ranged_value_to_percentage,
+)
+
+from . import (
+    ModernFormsDataUpdateCoordinator,
+    ModernFormsDeviceEntity,
+    modernforms_exception_handler,
+)
+from .const import (
+    ATTR_SLEEP_TIME,
+    CLEAR_TIMER,
+    DOMAIN,
+    OPT_ON,
+    OPT_SPEED,
+    SERVICE_CLEAR_FAN_SLEEP_TIMER,
+    SERVICE_SET_FAN_SLEEP_TIMER,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType, config_entry: ConfigEntry, async_add_entities: Callable
+) -> None:
+    """Set up a Modern Forms platform from config entry."""
+
+    coordinator: ModernFormsDataUpdateCoordinator = hass.data[DOMAIN][
+        config_entry.entry_id
+    ]
+
+    platform = entity_platform.async_get_current_platform()
+
+    platform.async_register_entity_service(
+        SERVICE_SET_FAN_SLEEP_TIMER,
+        {
+            vol.Required(ATTR_SLEEP_TIME): vol.All(
+                vol.Coerce(int), vol.Range(min=1, max=1440)
+            ),
+        },
+        "async_set_fan_sleep_timer",
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_CLEAR_FAN_SLEEP_TIMER,
+        {},
+        "async_clear_fan_sleep_timer",
+    )
+
+    update_func = partial(
+        async_update_fan, config_entry, coordinator, {}, async_add_entities
+    )
+    coordinator.async_add_listener(update_func)
+    update_func()
+
+
+class ModernFormsFanEntity(FanEntity, ModernFormsDeviceEntity):
+    """Defines a Modern Forms light."""
+
+    SPEED_RANGE = (1, 6)  # off is not included
+
+    def __init__(
+        self, entry_id: str, coordinator: ModernFormsDataUpdateCoordinator
+    ) -> None:
+        """Initialize Modern Forms light."""
+        super().__init__(
+            entry_id=entry_id,
+            coordinator=coordinator,
+            name=f"{coordinator.data.info.device_name} Fan",
+            icon="mdi:fan",
```

Comment:
You don't need to set an icon, the frontend already knows about fan entities. If set to `None`, frontend will pick the default one.

Reply:
Done